File: minjaeSong/LeetCode/LeetCode_690_ Employee Importance.java
// Definition for Employee.
// class Employee {
//     public int id;
//     public int importance;
//     public List<Integer> subordinates;
// };


class Solution {
    int sumValue = 0;
	
	void dfs(List<Employee> employees, int id, int value) {
		Employee emp = new Employee();
		for(Employee employee:employees) {
			if(id==employee.id) {
				emp = employee;
				sumValue += employee.importance;
			}				
		}
		if(emp.subordinates.size() == 0) {
			return;
		}
		for(int i=0;i<emp.subordinates.size();i++) {
			dfs(employees, emp.subordinates.get(i), value);
		}
		
	}
	
    public int getImportance(List<Employee> employees, int id) {
    	dfs(employees, id, 0);
        return sumValue;
    }
}

File: jeongbinlee/LeetCode/LeetCode_690_Employee Importance.cpp
class Solution {
private:
	int ret;
	vector<int> stack; 
	vector<Employee*> employeeList;
	
public:
	Solution()
	{
		ret = 0;
	}

	Employee getEmployeeById(int id)
	{
		Employee ret;

		for (int i = 0; i < employeeList.size(); i++)
		{
			if (employeeList.at(i)->id == id)
			{
				ret.id = employeeList.at(i)->id;
				ret.importance = employeeList.at(i)->importance;
				ret.subordinates = employeeList.at(i)->subordinates;
				break;
			}
		}

		return ret;
	}

	int getImportance(vector<Employee*> employees, int id) {
		int idx = 0;
		employeeList = employees;
		Employee who = getEmployeeById(id);
		ret += who.importance;

		bfs(who);

		for (int sum = 0; sum < stack.size(); sum++)
		{
			int imp = getEmployeeById(stack.at(sum)).importance;

			ret += imp;
		}
		
		return ret;
	}

	void bfs(Employee employee)
	{
		for (int sub = 0; sub < employee.subordinates.size(); sub++)
		{
			int subId = 0;
			if(employee.subordinates.size() > 0)
				subId = employee.subordinates.at(sub);
			if(!isExist(stack, subId))
				stack.push_back(subId);
			bfs(getEmployeeById(subId));
		}
	}

	bool isExist(vector<int> stack, int find)
	{
		for (int i = 0; i < stack.size(); i++)
			if (stack.at(i) == find)
				return true;

		return false;
	}
};
